crates/store/src/db/models/queries/transactions.rs

```diff
@@ -126,14 +174,211 @@ pub struct TransactionSummaryRowInsert {
+    let mut last_transaction_id: Option<Vec<u8>> = None;
+
+    loop {
+        let mut query =
+            SelectDsl::select(schema::transactions::table, TransactionRecordRaw::as_select())
+                .filter(schema::transactions::block_num.gt(block_range.start().to_raw_sql()))
+                .filter(schema::transactions::block_num.le(block_range.end().to_raw_sql()))
+                .filter(schema::transactions::account_id.eq_any(&desired_account_ids))
+                .into_boxed();
+
+        // Apply cursor-based pagination using the last seen (block_num, transaction_id)
+        if let (Some(last_block), Some(last_tx_id)) = (last_block_num, &last_transaction_id) {
+            query = query.filter(
+                schema::transactions::block_num
+                    .gt(last_block)
+                    .or(schema::transactions::block_num
+                        .eq(last_block)
+                        .and(schema::transactions::transaction_id.gt(last_tx_id))),
+            );
+        }
+
+        let chunk = query
+            .order((
+                schema::transactions::block_num.asc(),
+                schema::transactions::transaction_id.asc(),
+            ))
+            .limit(NUM_TXS_PER_CHUNK)
+            .load::<TransactionRecordRaw>(conn)
+            .map_err(DatabaseError::from)?;
+
+        // If no more data, we're done
+        if chunk.is_empty() {
+            break;
+        }
+
+        // Add transactions from this chunk one by one until we hit the limit
+        let mut added_from_chunk = 0;
+        for tx in chunk {
+            if total_size + tx.size_in_bytes <= MAX_PAYLOAD_BYTES {
+                total_size += tx.size_in_bytes;
+                // Update cursor position for next iteration
+                last_block_num = Some(tx.block_num);
+                last_transaction_id = Some(tx.transaction_id.clone());
+                all_transactions.push(tx);
+                added_from_chunk += 1;
+            } else {
+                // Can't fit this transaction, stop here
+                break;
+            }
+        }
+
+        // Break if chunk incomplete (size limit hit or data exhausted)
+        if added_from_chunk < NUM_TXS_PER_CHUNK {
+            break;
+        }
+    }
+
+    // Ensure block consistency: remove the last block if it's incomplete
+    // (we may have stopped loading mid-block due to size constraints)
+    let mut filtered_raw = all_transactions;
+
+    if total_size >= MAX_PAYLOAD_BYTES {
+        let last_idx = filtered_raw.len() - 1;
+        let last_block_num = filtered_raw[last_idx].block_num;
+        let prev_block_num = filtered_raw[last_idx - 1].block_num;
+
+        // If the last transaction is from the same block as the previous one,
+        // we may have stopped mid-block, so remove the entire last block
+        if last_block_num == prev_block_num {
+            filtered_raw.retain(|tx| tx.block_num != last_block_num);
+        }
```

Comment:
I'm not quite following this logic: why do we filter out the last block only if last block number is the same as the second to last block number? Shouldn't we do this regardless?

Also, could we not use the same general approach for filtering out the last block as we did in other places (e.g., for sync nullifiers) - e.g., something like:

```rust
if total_size >= MAX_PAYLOAD_BYTES {
    // TODO: filter out the last block
} else {
    Ok((vec_raw_try_into(filtered_raw)?, *block_range.end()))
}
```

Or would this not work for some reason?

Reply:
To be more clear, I ended up moving an approach more similar to the other paginated queries. With the difference than in the other we set a maximum number of rows, but here it is variable with the size. This way we ensure that the payload is always `< MAX_PAYLOAD_BYTES`.